libsolidity/codegen/ir/IRGeneratorForStatements.cpp

```diff
@@ -2514,22 +2514,34 @@ void IRGeneratorForStatements::appendExternalFunctionCall(
 		mstore(<pos>, <shl28>(<funSel>))
 		let <end> := <encodeArgs>(add(<pos>, 4) <argumentString>)
 
-		let <success> := <call>(<gas>, <address>, <?hasValue> <value>, </hasValue> <pos>, sub(<end>, <pos>), <pos>, <reservedReturnSize>)
+		let <success> := <call>(<gas>, <address>, <?hasValue> <value>, </hasValue> <pos>, sub(<end>, <pos>), <pos>, <staticReturndataSize>)
 		<?noTryCall>
 			if iszero(<success>) { <forwardingRevert>() }
 		</noTryCall>
 		<?+retVars> let <retVars> </+retVars>
 		if <success> {
-			<?dynamicReturnSize>
+			// Normally we don't need returndatacopy() because the CALL opcode will copy the return data automatically.
+			// However, when dynamicReturnSize is true, we don't know the size upfront.
+			// To handle this, ReturnInfo is supposed to set estimatedReturnSize to zero
+			// and we then need to copy the data out manually (based on returndatasize()).
+			<?isReturndataSizeDynamic>
 				// copy dynamic return data out
-				returndatacopy(<pos>, 0, returndatasize())
-			</dynamicReturnSize>
+				let <returnDataSizeVar> := returndatasize()
+				returndatacopy(<pos>, 0, <returnDataSizeVar>)
+			<!isReturndataSizeDynamic>
+				let <returnDataSizeVar> := <staticReturndataSize>
+				<?supportsReturnData>
+					if gt(<returnDataSizeVar>, returndatasize()) {
+						<returnDataSizeVar> := returndatasize()
+					}
+				</supportsReturnData>
+			</isReturndataSizeDynamic>
```

Comment:
Hmm... putting it all there in the generated code is actually a bit verbose. Let's spread it a bit and shorten it like this:

```c++
			<?isReturndataSizeDynamic>
				// explicit returndata copy needed: we don't know the size upfront so we passed zero to the <call>
				returndatacopy(...)
			<!isReturndataSizeDynamic>
				...
				</supportsReturnData>
				// no explicit returndata copy needed: we know the size upfront and we passed it to the <call>
			</isReturndataSizeDynamic>
```

Reply:
I don't think we should put comments there...
It won't look as convoluted anyway

Reply:
True, in Yul it'll be simpler. But what I really wanted was a comment at the C++ level in this particular spot. I think that what's happening here and why is not that obvious at first. Maybe we could instead do something like putting the text between `<?comment>` and `</comment>` and defining `comment` variable to be false so that it does not appear in the output?